aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp

```diff
@@ -955,6 +961,126 @@ void qtopk_kernel(Tensor& values,
   });
 }
 
+template <bool ReluFused>
+void q_batch_norm_kernel(
+    int N,
+    int C,
+    int HxW,
+    const int in_zero_point,
+    const int out_zero_point,
+    const uint8_t* X,
+    const float* alpha,
+    const float* beta,
+    uint8_t* Y) {
+#if defined(__AVX2__)
+  constexpr int kVLen = 8;
+  const int outer_size = N * HxW;
+  const __m256i min_v = _mm256_set1_epi32(std::numeric_limits<uint8_t>::min());
+  const __m256i max_v = _mm256_set1_epi32(std::numeric_limits<uint8_t>::max());
+  const __m256i shuffle_mask_v = _mm256_set_epi8(
+      0xff,
+      0xff,
+      0xff,
+      0xff,
+      0xff,
+      0xff,
+      0xff,
+      0xff,
+      0xff,
+      0xff,
+      0xff,
+      0xff,
+      0x0c,
+      0x08,
+      0x04,
+      0x00,
+      0xff,
+      0xff,
+      0xff,
+      0xff,
+      0xff,
+      0xff,
+      0xff,
+      0xff,
+      0xff,
+      0xff,
+      0xff,
+      0xff,
+      0x0c,
+      0x08,
+      0x04,
+      0x00);
+  const __m256i permute_mask_v =
+      _mm256_set_epi32(0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04, 0x00);
+  const __m256i in_zero_point_v = _mm256_set1_epi32(in_zero_point);
+  const __m256i out_zero_point_v = _mm256_set1_epi32(out_zero_point);
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+  for (int i = 0; i < outer_size; ++i) {
+    int n = C / kVLen * kVLen;
+    int r = C % kVLen;
+    const uint8_t* X_ptr = X + i * C;
+    uint8_t* Y_ptr = Y + i * C;
+    for (int j = 0; j < n; j += kVLen) {
```

Comment:
Can this instead be done 4x interleaved to ensure better pipeline occupancy? You can then replace most of these instructions with calls to Vec256<quint8>::quantize (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cpu/vec256/vec256_qint.h#L436)

That also cleans up the unnecessary shuffle instructions etc

Reply:
NB: I haven't compiled or run this so there might be some details off! :p

```
using Vec = Vec256<quint8>;

// Hoisted variables
auto in_zp_vec = Vec256<float>(static_cast<float>(in_zero_point));
auto fake_scale = Vec256<float>(1.0f);
auto scale_neg_zp_premul = fake_input_scale + in_zp.vec.neg();
auto out_zero_point_v = Vec(c10::quint8(out_zero_point));

for (int64_t j = 0; j < n; j += Vec::float_num_vecs()) {
  auto vals_q = Vec::loadu(X_ptr + j);
  // Fake scale of 1.0 here, should not affect performance (FMA in place of sub)
  auto vals_dq = vals_q.dequantize(fake_in_scale, in_zp_vec, scale_neg_zp_premul);
  for (int idx = 0; idx < vals_dq.size(); ++idx) {
    auto alpha_v = Vec256<float>(alpha + j * Vec::float_num_vecs() + idx);
    auto beta_v = Vec256<float>(beta + j * Vec::float_num_vecs() + idx);
    vals_dq[idx] = vec256::fmadd(vals_dq[idx], alpha_v, beta_v);
  }
  // Fake scale again
  auto outputs_q = Vec::quantize(vals_dq, /*output_scale=*/1.0f, out_zero_point, /*inv_output_scale=*/1.0f);
  if (ReluFused) {
    outputs_q = outputs_q.relu(out_zero_point_v);
  }
  outputs_q.store(Y_ptr + j);
}
```